nixos/modules/services/networking/nordvpn.nix

```
@@ -0,0 +1,79 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.nordvpn;
+in {
+  options = {
+    services.nordvpn = {
+      enable = mkEnableOption "nordvpn";
+
+      package = mkOption {
+        description = "NordVPN Package";
+        type = types.package;
+        default = pkgs.nordvpn;
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    boot.kernelModules = [ "tun" ];
+
+    networking.iproute2.enable = true;
+
+    users.groups.nordvpn = {};
+
+    systemd.sockets.nordvpnd = {
+      description = "NordVPN Daemon Socket";
+      partOf = [ "nordvpnd.service" ];
+      listenStreams = [ "/run/nordvpn/nordvpnd.sock" ];
+      wantedBy = [ "sockets.target" ];
+
+      socketConfig = {
+        NoDelay = "true";
+        SocketGroup = "nordvpn";
+        SocketMode = "0770";
+      };
+    };
+
+    systemd.services.nordvpnd = {
+      description = "NordVPN Daemon";
+      requires = [ "nordvpnd.socket" ];
+      after = [ "network-online.target" ];
+      wants = [ "network-online.target" ];
+
+      wantedBy = ["default.target"];
+
+      path = [
+        pkgs.iptables
+        pkgs.iproute2
+        pkgs.sysctl
+      ];
+
+      preStart = ''
+        cp -r ${cfg.package}/var/lib/nordvpn/* /var/lib/nordvpn/
+        cp ${pkgs.openvpn}/bin/openvpn /var/lib/nordvpn/openvpn
+      '';
+
+      serviceConfig = {
+        ExecStart = "${cfg.package}/bin/nordvpnd";
+        NonBlocking = "true";
+        KillMode = "process";
+        Restart = "on-failure";
+        RestartSec = "5";
+        Group = "nordvpn";
+      };
+    };
```

You can just use the upstream units that are provided with the package, for the most part.

```suggestion
    systemd.packages = [ cfg.package ];
    systemd.sockets.nordvpnd.wantedBy = [ "sockets.target" ];

    systemd.services.nordvpnd = {
      wantedBy = ["default.target"];

      path = with pkgs; [
        iptables
        iproute2
        sysctl
      ];

      preStart = ''
        cp -r ${cfg.package}/var/lib/nordvpn/* /var/lib/nordvpn/
        cp ${pkgs.openvpn}/bin/openvpn /var/lib/nordvpn/openvpn
      '';

      serviceConfig = {
        ExecStart = "${cfg.package}/bin/nordvpnd";
        StateDirectory = "nordvpn";
      };
    };
```

---

Additionally, can you please provide some context on the `preStart` and what exactly the files are used for, when they should be updated, etc...

---

the preStart process is needed to provide RSA key and server list for the nordvpn daemon and the openvpn copying is due to the openvpn that come with the package is using /sbin/ip which doesn't exist in nixos

---

it doesn't touch settings since that is generated by the daemon itself

---

I'm guessing we can't tell `nordvpn` to look in the nix store... Would a symlink be better? As it stands it looks like whenever the `nordvpn` package is updated it could cause some problems perhaps.

---

using something like lndir?

---

Cool 👍